source/MRMesh/MRObjectMeshHolder.cpp

```diff
@@ -315,6 +315,51 @@ void ObjectMeshHolder::copyTextureAndColors( const ObjectMeshHolder & src, const
     setUVCoords( std::move( uvCoords ) );
 }
 
+void ObjectMeshHolder::copyColors( const VisualObject& src, const VertMap& thisToSrc, const FaceMap& thisToSrcFaces )
+{
+    MR_TIMER
+
+    setColoringType( src.getColoringType() );
+
+    const auto& srcColorMap = src.getVertsColorMap();
+    if ( srcColorMap.empty() )
+        return;
+
+    VertColors colorMap;
+    colorMap.resizeNoInit( thisToSrc.size() );
+    ParallelFor( colorMap, [&] ( VertId id )
+    {
+        colorMap[id] = srcColorMap[thisToSrc[id]];
+    } );
+    setVertsColorMap( std::move( colorMap ) );
+
+    if ( !facesColorMap_.empty() && mesh_ )
+    {
+        auto& validFace = mesh_->topology.getValidFaces();
+        FaceColors faceColors;
+        faceColors.resize( validFace.size() );
+
+        Color color = facesColorMap_[thisToSrcFaces[validFace.backId()]];
+        bool differentColor = false;
+
+        for ( const auto& faceId : validFace )
+        {
+            auto& newColor = facesColorMap_[thisToSrcFaces[faceId]];
+            faceColors[faceId] = newColor;
+            if ( color != newColor )
+                differentColor = true;
+        }
```

Comment:
mb use ParallelFor with `tbb::enumerable_thread_specific` for same color check
check `thisToSrcFaces[faceId].valid()`